File: Python_Codes/9. Multithreading/WaitNotifyDemo.py
import threading

class WaitNotifyExample:
    def __init__(self):
        self.condition = threading.Condition()
        self.flag = False

    def producer(self):
        with self.condition:
            print("Producer is producing...")
            self.flag = True
            self.condition.notify()

    def consumer(self):
        with self.condition:
            while not self.flag:
                self.condition.wait()
            print("Consumer consumed.")

def execute_wait_notify_example():
    example = WaitNotifyExample()

    producer_thread = threading.Thread(target=example.producer)
    consumer_thread = threading.Thread(target=example.consumer)

    consumer_thread.start()
    producer_thread.start()

    producer_thread.join()
    consumer_thread.join()

if __name__ == "__main__":
    execute_wait_notify_example()

File: Java_Codes/9. Multithreading/WaitNotifyDemo.java
//Coordinating a producer-consumer problem in messaging queues.
class WaitNotifyExample {
    private final Object lock = new Object();
    private boolean flag = false;

    public void producer() throws InterruptedException {
        synchronized (lock) {
            System.out.println("Producer is producing...");
            flag = true;
            lock.notify();
        }
    }

    public void consumer() throws InterruptedException {
        synchronized (lock) {
            while (!flag) {
                lock.wait();
            }
            System.out.println("Consumer consumed.");
        }
    }
}

public class WaitNotifyDemo {
    public static void executeWaitNotifyExample() {
        WaitNotifyExample example = new WaitNotifyExample();

        Thread producerThread = new Thread(() -> {
            try {
                example.producer();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        Thread consumerThread = new Thread(() -> {
            try {
                example.consumer();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        consumerThread.start();
        producerThread.start();
    }

    public static void main(String[] args) {
        executeWaitNotifyExample();
    }
}

